### ISSUE TITLE
Severe performance degradation in reading logs since 8.4.0

### ISSUE THREAD

**Issue by piotrp** [elastic/beats#33749]
Reading of log files from disk got a lot slower since Filebeat 8.4.0, and 8.5.0 made it even worse.

Test configuration with `log` input:
```yaml
filebeat.inputs:
  - type: log
    paths:
      - test-file.log
    encoding: utf-8
    close_eof: true

output:
  console:
    codec.format.string: "."
```

Test configuration with `filestream` input:
```yaml
filebeat.inputs:
  - type: filestream
    id: test-logs
    paths:
      - test-file.log
    encoding: utf-8
    close.reader.on_eof: true

output:
  console:
    codec.format.string: "."
```

Run it using:
```
seq 1000000 > test-file.log
# variant 1: get logs with timestamps in ./tmp
rm -rf ./tmp; ./filebeat-8.3.3 --path.home . --path.config . --path.logs tmp --path.data tmp -c fb-test-log.yml > /dev/null
rm -rf ./tmp; ./filebeat-8.4.3 --path.home . --path.config . --path.logs tmp --path.data tmp -c fb-test-log.yml > /dev/null
# variant 2: pipe to pv to get log lines/s
rm -rf ./tmp; ./filebeat-8.3.3 --path.home . --path.config . --path.logs tmp --path.data tmp -c fb-test-log.yml | pv -Warl >/dev/null
rm -rf ./tmp; ./filebeat-8.4.3 --path.home . --path.config . --path.logs tmp --path.data tmp -c fb-test-log.yml | pv -Warl >/dev/null
```

Filebeat 8.3.3 can process the `log` test case below in 9,175s. Filebeat 8.4.3 needs 14,966s.

When checking speed with `pv` I get consistently:
| Filebeat | `log` | `filestream` | notes
|-|-|-|-
| 7.11.2 | 175 k/s | \- |
| 8.3.3 | 105-110 k/s | 152 k/s |
| 8.4.3 | 66,5 k/s | 90 k/s |
| 8.5.1 | 54 k/s | 66,5 k/s |
| 8.5.2 | 87 k/s | 100 k/s | #33620 backported

_(11.02.2023: added results for 8.5.2)_

Test platform:
* Ubuntu 20.04.5 LTS / 5.4.0-84-generic (amd64)
* CPU E5-2667 v4 @ 3.20GHz

A test on real JSON logs with Logstash output also shows similar results (same configuration, different Filebeat versions):
![obraz](https://user-images.githubusercontent.com/1409807/203159635-765c41b5-e2d2-4e09-be31-b324f599c6bc.png)

**Comment by piotrp** [comment]
Looks like it got better in 8.5.2, probably thanks to #33620 (#33624):

| Filebeat | `log` | `filestream`
|-|-|-
| 8.5.2 | 87 k/s | 100 k/s

Still not 8.3 level of performance, but at least it's no longer two times slower.

**Comment by alexsapran** [comment]
Thanks, @piotrp for the report and the information you provided. 

After an internal investigation indeed the issue is fixed at https://github.com/elastic/beats/pull/33620. 

The mentioned PR changes the `DEV=true` for the released artifacts which disabled the optimizations that caused the performance decrees.
We are working on establishing continuous performance testing to catch similar issues in the future.

I am going to close the issue, as there is no further investigation required.

**Comment by hostalp** [comment]
#33620 isn't it. It improves the performance a bit, but it's still far from what it used to be. Just see one of the older comments: https://github.com/elastic/beats/issues/33749#issuecomment-1325149742

**Comment by piotrp** [comment]
I agree, this doesn't fix slowdown from 8.4 (and 8.x series in general). New tests, each re-run three times to verify that results are stable:

| Filebeat | `log` | `filestream` | notes
|-|-|-|-
| 8.3.3 | 100-105 k/s | 152 k/s  |
| 8.4.3 | 66 k/s | 89 k/s  |
| 8.5.2 | 85 k/s | 100 k/s | with #33624 (backport of #33620)
| 8.6.1 | 87 k/s | 100 k/s | with #33620

**Comment by cmacknz** [comment]
Let's reopen this then until we can track down what the difference between 8.3.3 and 8.6.1 is. There were no major changes in that period to obviously blame for this (besides releasing with the wrong build flags).
